// @easypost/build/src/webpack/rules/js.js
import {
  DEFAULT_NODE_MODULES_TO_COMPILE,
  generateExcludes,
} from '../helpers/externals';

export const TARGET_BROWSERS = {
  browsers: ['last 2 versions', 'safari >= 7'],
};

export const TARGET_NODE_8 = {
  node: '8.11',
};


export default (targets = TARGET_BROWSERS, opts = {}) => {
  const {
    nodeModulesToCompile = DEFAULT_NODE_MODULES_TO_COMPILE,
    ...options
  } = opts;

  const excludes = generateExcludes(nodeModulesToCompile);

  return {
    test: /\.jsx?$/,
    loader: 'babel-loader',
    exclude: excludes,
    options: {
      babelrc: false,

      presets: [[
        '@babel/preset-env', {
          targets,
        },
      ], [
        '@babel/preset-react',
      ]],

      plugins: [
        '@babel/plugin-proposal-export-namespace-from',
        ['@babel/plugin-proposal-class-properties', {
          loose: false,
        }],
        '@babel/plugin-proposal-export-default-from',
        '@babel/plugin-syntax-export-extensions',
        'babel-plugin-transform-export-extensions',
        '@babel/plugin-proposal-optional-chaining',
        '@babel/plugin-proposal-nullish-coalescing-operator',
      ],

      ...options,
    },
  };
};
